### PR TITLE
perf(core): fit gridded Pearson Type III parameters once per spatial block (#940)

### LINKED ISSUE #940: VECTORIZE-893: Vectorize the Pearson Type III per-series fit across grid cells
Part of the #893 performance epic. Follow-up from the spatial `apply_ufunc` conversion in #923.

## Problem

#923 converts the fitting-based indices (SPI, SPEI) to spatial `apply_ufunc` blocks, so
gridded gamma fitting/transform now runs once per cell set with no Python loop over grid
cells. The Pearson Type III path still runs one Python call per cell:

- `compute.pearson_parameters` loops over the calendar time steps and calls
  `calculate_time_step_params` per series, which fits `lmoments.fit` to one 1-D sample.
- `compute._check_goodness_of_fit_pearson` performs the same per-series K-S check.
- `indices.spi` / `indices.spei` therefore re-enter the single-series kernel per cell for
  `Distribution.pearson` (`indices._apply_per_cell`), which is correct but still a
  per-cell Python loop.

SPI/SPEI default to gamma, so the #923 acceptance criterion (no loop for gridded SPI) is
met, but a gridded Pearson run is unchanged from before.

## Tasks

- [ ] Decide whether L-moments can be evaluated across a cell axis as arrays (the
      probability-of-zero and sample-size guards are already per-step vector quantities)
      or whether the K-S check is the only part worth vectorizing.
- [ ] Record the decision on this issue before implementing.
- [ ] Add the cell-axis path to `compute.pearson_parameters` / `_pearson_fit` and remove
      `indices._apply_per_cell`, or rename it if another index still needs it.
- [ ] Keep the NumPy API contract for `compute.pearson_parameters`,
      `indices.spi`, and `indices.spei` unchanged.

## Acceptance criteria

- [ ] Gridded SPI with `Distribution.pearson` runs with no Python-level loop over grid cells
      (extend the no-loop guard added in #923 to cover both distributions).
- [ ] Results match the single-series path within the tolerance asserted in #923's
      equivalence tests, bit-for-bit where feasible.
- [ ] `uv run ruff check src/ tests/`, `uv run ruff format --check src/ tests/`,
      `uv run mypy src/`, and `uv run pytest` pass.

## Related

- #923 — the spatial `apply_ufunc` conversion that defers this path.
- #922 / PR #936 — the inventory that lists the per-cell invocation sites.
- #929 (baselines), #930 (equivalence tests).

### PR DESCRIPTION
**Pull request by monocongo**
Closes #940.

## Summary

- Adds a cell-axis L-moment fit (`lmoments.fit_spatial`) and routes `compute.pearson_parameters` through it when the calibration values are a folded `(years, time_steps, *cells)` block, so gridded SPI/SPEI with `Distribution.pearson` fits every cell in one pass.
- Keeps the single-series addition order per cell (the same rank loop with every cell evaluated at once), so parameter values match the per-series fits; only `scale` differs in the last ulp from `np.exp`/`gammaln` against `math.exp`/`lgamma`.
- Adds `_check_goodness_of_fit_pearson_spatial`, mirroring the gamma spatial check: one vectorized D statistic per time step across every cell, SciPy only for candidates at the critical value.
- Removes `indices._apply_per_cell`; both call sites (SPI and SPEI Pearson branches) are gone.
- Re-homes the caller-supplied fitting-parameter cell validation into `fit_and_standardize`, before the Pearson-to-gamma fallback try, so a mismatched parameter array still raises instead of silently falling back to gamma.
- Fixes the spatial windowed sum (`sum_to_scale`) so a masked cell is filled with NaN instead of reading under the mask; this is what the removed per-cell Pearson path had been compensating for.

The decision record is on the issue.

## Behavior notes

- The Pearson-to-gamma fallback now applies to the whole block rather than per cell, the same consequence the gamma spatial conversion in #923 accepted; noted in ADR-0008.
- The decision record on #940 records this; the alternative (per-cell fallback) would defeat the vectorization.

## Validation

- `uv run ruff check src/ tests/` — clean
- `uv run ruff format --check src/ tests/` — clean
- `uv run mypy src/` — clean
- `uv run pytest` — 2082 passed, 423 deselected
- `uv run pytest -m validation` — 378 passed, 3 skipped
- `uv run scripts/generate_llms_txt.py` regenerated `llms.txt` / `llms-full.txt` for the `docs/xarray_compatibility.md` edit; `tests/test_review_scripts.py` passes.

## Summary by Sourcery

Fit Pearson Type III parameters and validate goodness of fit once per gridded spatial block instead of once per cell.

New Features:
- Vectorize Pearson Type III L-moment fitting and goodness-of-fit checks across gridded spatial blocks for SPI and SPEI.
- Support broadcasting and validation of period-only or cell-aware precomputed Pearson parameters for spatial inputs.

Bug Fixes:
- Preserve missing values during spatial windowed summation by converting masked entries to NaN.
- Reject mismatched spatial fitting parameters before Pearson fallback can incorrectly treat them as fit failures.

Enhancements:
- Remove the per-cell Pearson execution path while retaining pointwise-equivalent results and whole-block fallback behavior.
- Share the vectorized spatial goodness-of-fit machinery between gamma and Pearson checks.

Documentation:
- Update spatial-kernel documentation and ADR guidance to describe gridded Pearson fitting, memory considerations, and whole-block fallback semantics.

Tests:
- Add coverage for vectorized SPI/SPEI Pearson call counts, pointwise equivalence, invalid samples, parameter broadcasting and validation, fallback behavior, and goodness-of-fit parity.

## Summary by CodeRabbit

* **New Features**
  * Added one-pass Pearson Type III fitting and goodness-of-fit processing for gridded SPI/SPEI data.
  * Pearson parameters now support consistent broadcasting across all spatial cells.
  * Invalid spatial fits receive a block-level gamma fallback.

* **Bug Fixes**
  * Improved handling of masked and incomplete spatial inputs.
  * Added validation for mismatched precomputed parameters and spatial dimensions.

* **Documentation**
  * Updated compatibility guides and architecture documentation to reflect the spatial Pearson workflow.

## Scope note

The fit, transform, and vectorized goodness-of-fit D statistic run once per block. SciPy's exact K-S p-value is still evaluated per flagged (time step, cell) candidate, matching the per-cell check's count of exact evaluations; only the flagging is vectorized.

### CODE REVIEW

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** [PR #982 — perf(core): fit gridded Pearson Type III parameters once per spatial block](https://github.com/monocongo/climate_indices/pull/982) (`perf/940-pearson-spatial-fit`, reviewed through `95d7cc8b`, fixes verified at `32d2ba37`)
- **Lanes run:** 4 — correctness/edge cases, dataflow/shape contracts, performance, tests/contracts
- **Validation:** ruff check, ruff format --check, mypy, `pytest` (2096 passed, 423 deselected), `pytest -m validation` (378 passed, 3 skipped), `pytest tests/test_review_scripts.py` (13 passed); fix-pass findings reproduced and re-verified directly
- **Fix commits:** [32d2ba37 — fix(core): align spatial Pearson fits with the single-series fallback](https://github.com/monocongo/climate_indices/pull/982/commits/32d2ba37)

### Findings and fix status

#### [P2] Rejected L-moments left the probability of zero and the failure count unset — Fixed
- **Location:** `src/climate_indices/lmoments.py:129-131`, `src/climate_indices/compute.py:446-451`
- **Origin:** Correctness/edge cases
- **Problem:** `fit_spatial` returned the coarse validity mask (`sample count >= 4` and non-zero second L-moment), while `_estimate_pearson3_parameters_spatial` refined it locally (`|L-skew| < 1`) without returning it. A cell rejected by the refined check (e.g. a five-value sample `[0, 20, 20, 20, 20]`, `|L-skew| = 1`) kept a non-zero probability-of-zero and was not counted as a failed fit, unlike the single-series `except DistributionFittingError` branch.
- **Applied fix:** `_estimate_pearson3_parameters_spatial` now returns the refined mask and `fit_spatial` propagates it; verified `p0 = 0` and `failed = 1` for the reproduced sample. Regression test `test_pearson_parameters_spatial_matches_pointwise_for_invalid_samples` compares every parameter against the per-cell fit, including that cell. Commit `32d2ba37`.

#### [P2] Scalar and cell-axis `scale` diverged for near-symmetric samples — Fixed
- **Location:** `src/climate_indices/lmoments.py:106`
- **Origin:** Correctness/edge cases
- **Problem:** the scalar fit used `math.lgamma` and the cell-axis fit `scipy.special.gammaln`; at L-skew `2e-6` (alpha ≈ 3e10) the two round the large-value subtraction differently and `scale` differed by `1.2e-4` relative, above the equivalence tolerance the tests assert.
- **Applied fix:** the scalar fit uses `special.gammaln` too; measured relative difference is now 0 across the cutoff band. Commit `32d2ba37`.

#### [P2] Goodness-of-fit candidate loop recomputed the block CDF per flagged cell — Fixed
- **Location:** `src/climate_indices/compute.py:1283-1290`
- **Origin:** Performance
- **Problem:** the vectorized CDF evaluated for every cell was discarded, and the exact-p-value branch recomputed `pearson3.cdf` per candidate. Measured on a synthetic poorly-fitting #893-shaped block (38×87, 40 years): 9.0 s end to end, of which 7.55 s was the candidate recompute.
- **Applied fix:** the candidate loop reuses the matching slice of the precomputed `cdf_values`; measured 16% lower end-to-end time with identical flagged counts. The gamma spatial check keeps the same pre-existing pattern, unchanged here. Commit `32d2ba37`.

#### [P3] Masked values were still read from under the mask in the series sum path — Fixed
- **Location:** `src/climate_indices/compute.py:266-287`
- **Origin:** Dataflow/shape contracts
- **Problem:** the windowed spatial branch filled masked entries with NaN, but the 1-D/2-D `np.convolve` branch still summed the data under the mask, so the same masked input gave NaN on one path and silent values on the other.
- **Applied fix:** the masked-to-NaN fill now runs before both branches. Commit `32d2ba37`.

#### [P3] Whole-block Pearson-to-gamma fallback was documented but untested — Fixed
- **Location:** `src/climate_indices/compute.py:1829-1832`, `docs/adr/0008-spatial-block-declaration.md`
- **Origin:** Tests/contracts
- **Problem:** the block-level fallback decision (an exception, or more than 20% missing after the transform, re-fits the whole block rather than one cell) had no regression pin, and the ADR wording implied every failed fit triggered it.
- **Applied fix:** `test_pearson_failure_falls_back_to_gamma_for_the_whole_block` monkeypatches a raised fit and asserts the result equals the gamma run; the `fit_and_standardize` docstring and the ADR now state that the decision is made once per block on the block-wide missing share. Commit `32d2ba37`.

#### [P3] The Pearson no-loop guard covered only the SPI branch — Fixed
- **Location:** `tests/test_spatial_kernel.py`
- **Origin:** Tests/contracts
- **Problem:** issue #940 asks to extend the no-loop guard to both distributions; only SPI-Pearson pinned one fit and one transform call, so a per-cell regression in the SPEI branch would have passed.
- **Applied fix:** `test_spei_pearson_fits_once_for_gridded_input` mirrors the guard through the SPEI adapter. Commit `32d2ba37`.

#### [P3] Pearson spatial guards had no zero-heavy or invalid-sample coverage — Fixed
- **Location:** `tests/test_spatial_kernel.py`
- **Origin:** Tests/contracts
- **Problem:** every Pearson spatial fixture used complete, positive data, so the minimum-non-zero guard, the probability-of-zero branch, and the invalid-L-moment mask never fired.
- **Applied fix:** the invalid-sample parity test above exercises the rejected-cell fallback. Commit `32d2ba37`.

#### [P4] Stale per-cell cite of #940 in `CONTEXT.md` — Fixed
- **Location:** `src/climate_indices/CONTEXT.md:139`
- **Origin:** Dataflow/tests
- **Problem:** the Per-Cell Path list still named #940, which this PR resolves; `xarray_compatibility.md`, `benchmarks/README.md`, and ADR-0008 were updated but this cite was missed.
- **Applied fix:** #940 dropped from the list. Commit `32d2ba37`.

#### [P4] Guaranteed full-block copies in the spatial goodness-of-fit path — Fixed
- **Location:** `src/climate_indices/compute.py:1254-1259`
- **Origin:** Performance
- **Problem:** `sorted_values` is always float64 (the `np.inf` substitution promotes it), so `.astype(float)` copied the whole block on every call; the parameter casts did the same for float64 input.
- **Applied fix:** `np.asarray(..., dtype=float)`, which is a no-op for matching dtypes. Commit `32d2ba37`.

#### [P4] Pearson CDF memory multiplier absent from the chunk guidance — Fixed
- **Location:** `docs/adr/0008-spatial-block-declaration.md`
- **Origin:** Performance
- **Problem:** the scipy `pearson3.cdf` machinery holds a few extra block-sized temporaries, so a Pearson block peaks higher than the same block on gamma; the documented chunk budget did not mention it.
- **Applied fix:** the ADR's memory paragraph now states the multiplier. Commit `32d2ba37`.

#### [P4] PR description overstated the wall-clock win — Fixed
- **Location:** PR #982 description
- **Origin:** Performance
- **Problem:** "one pass" is accurate for the fit and D statistic, but the exact SciPy p-value still runs per flagged candidate, so end-to-end gains on poorly fitted data are smaller than the description implied.
- **Applied fix:** a scope note now states that the exact K-S p-value remains per flagged candidate.

### Remaining blockers
None. All verified findings are fixed and the full validation gate passes.

File: docs/adr/0008-spatial-block-declaration.md [hunk 1]
```diff
@@ -53,10 +53,11 @@ continental grid. Dask still requires the time dimension in a single chunk, for
...
 window.
 
-The Pearson Type III branch keeps [the per-cell path](../xarray_compatibility.md); its L-moment fit
-is per series, so `Distribution.pearson` still re-enters the single-series kernel once per cell
-(issue #940). EDDI and percentage-of-normal have no cell axis in their kernels yet (#942), and
-Palmer has no adapter layer at all (#937).
+The Pearson Type III branch fits its L-moment parameters across the cell axis as well
+([#940](https://github.com/monocongo/climate_indices/issues/940)), so `Distribution.pearson` no
+longer re-enters the single-series kernel once per cell; like gamma, a failed fit falls back to
+gamma for the whole block rather than per cell. EDDI and percentage-of-normal have no cell axis in
```
[the remark this answers is not in the document]
Comment (monocongo) [reply docs/adr/0008-spatial-block-declaration.md]: Fixed in 28e5e403: reworded the ADR to scope the whole-block gamma fallback to `indices.spi` (`fallback_to_gamma=True`), and note that `indices.spei` (`fallback_to_gamma=False`) propagates Pearson fitting failures instead of falling back.

File: src/climate_indices/lmoments.py [hunk 2]
```diff
@@ -108,6 +116,126 @@ def _estimate_pearson3_parameters(lmoments: np.ndarray) -> dict[str, float]:
...
+    Returns the L-Moments fits (loc, scale, skew) for every cell of an array whose
+    first axis is the sample axis.
+
+    Cell-axis counterpart of :func:`fit`: instead of raising on the first cell whose
+    sample is too short or whose L-moments are invalid, the invalid cells are marked
+    in the returned validity mask so the caller can apply its own fallback.
+
+    :param values: array of samples with shape (samples, *cells)
+    :return: tuple of (loc, scale, skew, valid), each array shaped like values.shape[1:]
+    """
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/lmoments.py]: Fixed in 28e5e403: converted `fit_spatial`'s docstring to Google-style `Args:`/`Returns:` sections per CONTRIBUTING.md, same parameter/shape/return content.

File: src/climate_indices/lmoments.py [hunk 3]
```diff
@@ -108,6 +116,126 @@ def _estimate_pearson3_parameters(lmoments: np.ndarray) -> dict[str, float]:
...
+    lmoments, valid = _estimate_lmoments_spatial(values)
+    locs, scales, skews = _estimate_pearson3_parameters_spatial(lmoments, valid)
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/lmoments.py]: Already fixed on this branch in 32d2ba37: `_estimate_pearson3_parameters_spatial` now refines `valid` with the same `lmoments[1] > 0` / `abs(lmoments[2]) < 1.0` predicate and returns it, and `fit_spatial` propagates that refined mask instead of the pre-refinement one. Covered by `test_pearson_parameters_spatial_matches_pointwise_for_invalid_samples` in `tests/test_spatial_kernel.py`, which reproduces this exact `abs(L-skew) == 1` case.

File: src/climate_indices/lmoments.py [hunk 4]
```diff
@@ -108,6 +116,126 @@ def _estimate_pearson3_parameters(lmoments: np.ndarray) -> dict[str, float]:
...
+
+    The per-series accumulator is a sequential loop over the sorted sample ranks; the
+    same loop runs here with every cell evaluated at once, so each cell's additions
+    happen in the same order as the single-series fit.
+
+    :param values: array of samples with shape (samples, *cells)
+    :return: tuple of (lmoments, valid) with lmoments shaped (3, *cells) and valid
+        shaped (*cells)
+    """
+    values = np.asarray(values, dtype=float)
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/lmoments.py]: Fixed in 28e5e403: `_estimate_lmoments_spatial` now detects a `MaskedArray` and fills masked entries with NaN before sorting, instead of reading the backing values under the mask via `np.asarray`. Reproduced the bug directly against `lmoments.fit_spatial` (a masked outlier previously skewed `loc`), confirmed the fix, and the full suite (2096 tests) passes.

File: src/climate_indices/compute.py [hunk 5]
```diff
@@ -419,6 +423,41 @@ def calculate_time_step_params(time_step_values: np.ndarray) -> tuple[float, flo
...
+    number_of_non_missing = np.count_nonzero(~np.isnan(calibration_values), axis=0)
+    non_zero_count = number_of_non_missing - number_of_zeros
+    valid = valid & (non_zero_count >= MIN_NON_ZERO_VALUES_FOR_PEARSON)
+
+    with np.errstate(divide="ignore", invalid="ignore"):
+        probabilities_of_zero = np.where(number_of_zeros > 0, number_of_zeros / number_of_non_missing, 0.0)
+
+    failed_fitting_count = int(np.count_nonzero(~valid))
+    return (
+        np.where(valid, probabilities_of_zero, 0.0),
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/compute.py:456]: Fixed by commit `32d2ba37` (returns the refined L-moment validity mask from `fit_spatial`, so a cell whose L-moments are rejected zeroes its probability of zero and counts as a failed fit exactly as the single-series branch does) and `28e5e403` (scopes the ADR wording to `indices.spi`, which is the only caller with `fallback_to_gamma=True`, and to whole-block granularity, matching actual behavior). Both were already on this branch before this review round. The remaining block-wide (not per-cell) NaN-percentage trigger for the gamma fallback is the documented, intentional design in ADR-0008 for the vectorized path, not an oversight.

File: src/climate_indices/compute.py [hunk 6]
```diff
@@ -636,6 +695,59 @@ def _pearson_fit(
...
+def _validate_pearson_parameter_cells(
+    values: np.ndarray, named_parameters: tuple[tuple[str, np.ndarray | None], ...]
+) -> None:
+    """Reject pre-computed Pearson parameters whose cell axes do not match a block."""
+    cells = values.shape[2:]
+    for name, parameter in named_parameters:
+        if parameter is None:
+            continue
+        parameter = np.asarray(parameter)
+        if parameter.ndim > 1 and parameter.shape[1:] != cells:
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/compute.py]: Fixed in commit `44a44a75`: `_validate_pearson_parameter_cells` now also checks the supplied parameter's leading period-axis length against the block's period count, not just its trailing cell axes. A `(1, *cells)` array (or any other period-length mismatch) now raises immediately, before `fallback_to_gamma` can catch it, instead of silently broadcasting one period across every month/day or being swallowed into a gamma fallback. Covered by the new `test_period_mismatched_parameter_cells_raise` regression test.

File: src/climate_indices/compute.py [hunk 7]
```diff
@@ -1083,6 +1209,114 @@ def _check_goodness_of_fit_pearson(
...
+    for valid_count in np.unique(valid_counts):
+        if valid_count > 0:
+            critical_values[valid_counts == valid_count] = _ks_critical_value(int(valid_count))
+    critical_tolerance = 0.0
+    if np.issubdtype(calibration_values.dtype, np.floating):
+        critical_tolerance = float(np.finfo(calibration_values.dtype).eps)
+
+    candidates = np.nonzero(d_statistics >= critical_values - critical_tolerance)
+    poor_fits: list[tuple[int, float]] = []
+    for candidate in zip(*candidates, strict=True):
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/compute.py]: Not changed further, kept as-is. The redundant CDF recomputation this finding calls out was already removed in `32d2ba37`/`97ed5166`: `_spatial_poor_fits` now reuses the already-vectorized `cdf_values` instead of recomputing `pearson3.cdf` per candidate. What remains — one `scipy.stats.kstest` call per flagged candidate to get an exact p-value — is the same pattern the pre-existing single-series check already uses (see `_ks_poor_fit_p_value`, kept deliberately non-vectorized "to match `scipy.stats.kstest` at the threshold, including its version-specific dtype handling"). Reimplementing that p-value computation in a vectorized way risks diverging from SciPy's own version-specific KS numerics, which this PR has otherwise gone out of its way to match bit-for-bit against the per-series path. Given the Minor/advisory severity and that this only engages for candidates already past the vectorized D-statistic screen, leaving it consistent with the existing per-series design is the better trade-off here.
